apps/aecore/src/aec_peer_connection.erl

```diff
@@ -0,0 +1,825 @@
+        , get_header_by_height/2
+        , get_mempool/1
+        , get_n_successors/3
+        , ping/1
+        , send_block/2
+        , send_tx/2
+
+        , stop/1
+        ]).
+
+-behaviour(gen_server).
+
+-import(aeu_debug, [pp/1]).
+
+-include("common.hrl").
+-include("blocks.hrl").
+
+-define(MSG_FRAGMENT, 0).
+-define(MSG_PING, 1).
+-define(MSG_PING_RSP, 2).
+-define(MSG_GET_HEADER_BY_HASH, 3).
+-define(MSG_GET_HEADER_BY_HASH_RSP, 4).
+-define(MSG_GET_N_SUCCESSORS, 5).
+-define(MSG_GET_N_SUCCESSORS_RSP, 6).
+-define(MSG_GET_BLOCK, 7).
+-define(MSG_GET_BLOCK_RSP, 8).
+-define(MSG_SEND_TX, 9).
+-define(MSG_SEND_TX_RSP, 10).
+-define(MSG_SEND_BLOCK, 11).
+-define(MSG_SEND_BLOCK_RSP, 12).
+-define(MSG_GET_MEMPOOL, 13).
+-define(MSG_GET_MEMPOOL_RSP, 14).
+-define(MSG_GET_HEADER_BY_HEIGHT, 15).
+-define(MSG_GET_HEADER_BY_HEIGHT_RSP, 16).
+
+-define(FIRST_PING_TIMEOUT, 30000).
+-define(NOISE_HS_TIMEOUT, 5000).
+
+%% gen_server callbacks
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
+         terminate/2, code_change/3]).
+
+%% -- API --------------------------------------------------------------------
+
+connect(Options) ->
+    lager:debug("New peer connection to ~p", [Options]),
+    gen_server:start_link(?MODULE, [Options#{ role => initiator }], []).
+
+accept(TcpSock, Options) ->
+    {ok, {Host, _Port}} = inet:peername(TcpSock),
+    case gen_server:start_link(?MODULE, [Options#{ tcp_sock => TcpSock, role => responder,
+                                                   host => inet_parse:ntoa(Host)}], []) of
+        {ok, Pid} ->
+            gen_tcp:controlling_process(TcpSock, Pid),
+            gen_server:cast(Pid, give_tcp_socket_ownership),
+            {ok, Pid};
+        Res ->
+            Res
+    end.
+
+retry(PeerCon) ->
+    gen_server:cast(PeerCon, retry).
+
+ping(PeerId) ->
+    call(PeerId, ping).
+
+get_header_by_hash(PeerId, Hash) ->
+    call(PeerId, {get_header_by_hash, Hash}).
+
+get_header_by_height(PeerId, Hash) ->
+    call(PeerId, {get_header_by_height, Hash}).
+
+get_n_successors(PeerId, Hash, N) ->
+    call(PeerId, {get_n_successors, Hash, N}).
+
+get_block(PeerId, Hash) ->
+    call(PeerId, {get_block, Hash}).
+
+get_mempool(PeerId) ->
+    call(PeerId, get_mempool).
+
+send_tx(PeerId, Tx) ->
+    call(PeerId, {send_tx, Tx}).
+
+send_block(PeerId, Tx) ->
+    call(PeerId, {send_block, Tx}).
+
+stop(PeerCon) ->
+    gen_server:cast(PeerCon, stop).
+
+
+call(PeerCon, Call) when is_pid(PeerCon) ->
+    gen_server:call(PeerCon, Call);
+call(PeerId, Call) when is_binary(PeerId) ->
+    case aec_peers:get_connection(PeerId) of
+        {ok, PeerCon}    ->
+            case PeerCon of
+                {connected, Pid} -> call(Pid, Call);
+                _ -> {error, no_connection}
+            end;
+        Err = {error, _} ->
+            Err
+    end.
+
+
+%% -- gen_server callbacks ---------------------------------------------------
+init([Opts]) ->
+    Version = <<?PROTOCOL_VERSION:16>>,
```

Comment:
```
6> <<16#0000:16>>.
<<0,0>>
7> <<16#1000:16>>.
<<16,0>>
8> <<16#10000:16>>.
<<0,0>>
```

It looks like this shall cater for 8 bytes https://github.com/aeternity/epoch/blob/95b2d26ca66dc3c4f59b7bef2134afd82b1cee6c/apps/aecore/src/aec_headers.erl#L109 https://github.com/aeternity/protocol/pull/56/files#diff-3976273847c0bfced542c67100a7990bR25